### PR TITLE
chore: export pipeline related metrics

### PR DESCRIPTION
**Pull request by romange**
Export in /metrics
1. Total pipeline queue length
2. Total pipeline commands
3. Total pipelined duration

### CODE REVIEW

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -1272,12 +1259,16 @@ void Connection::SquashPipeline(facade::SinkReplyBuilder* builder) {
   cc_->async_dispatch = false;
 
   auto it = dispatch_q_.begin();
-  while (it->IsIntrusive())  // Skip all newly received intrusive messages
+  while (it->IsControl())  // Skip all newly received intrusive messages
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:1262]: Until we come up with an intrusive data message type 🙂

**Review by dranikpg — DISMISSED** [review]

File: src/facade/facade.cc [hunk 2]
```diff
@@ -20,12 +20,13 @@ constexpr size_t kSizeConnStats = sizeof(ConnectionStats);
...
 ConnectionStats& ConnectionStats::operator+=(const ConnectionStats& o) {
   // To break this code deliberately if we add/remove a field to this struct.
-  static_assert(kSizeConnStats == 112u);
+  static_assert(kSizeConnStats == 120u);
 
   ADD(read_buf_capacity);
   ADD(dispatch_queue_entries);
   ADD(dispatch_queue_bytes);
   ADD(dispatch_queue_subscriber_bytes);
+  ADD(dispatch_queue_max_len);
```
Comment (adiholden) [inline src/facade/facade.cc:29]: here by ADD you will get the sum of all queue max len, is this what you want?

File: src/facade/dragonfly_connection.cc [hunk 3]
```diff
@@ -1291,11 +1282,15 @@ void Connection::ClearPipelinedMessages() {
...
     FiberAtomicGuard guard;  // don't suspend when concluding to avoid getting new messages
-    if (msg.IsIntrusive())
+    if (msg.IsControl())
       visit(dispatch_op, msg.handle);  // to not miss checkpoints
     RecycleMessage(std::move(msg));
   }
 
+  if (dispatch_q_.size() == stats_->dispatch_queue_max_len) {
+    // conservative reset that will be corrected upon next insertion.
+    stats_->dispatch_queue_max_len = 0;
```
Comment (adiholden) [inline src/facade/dragonfly_connection.cc:1292]: why do you call it dispatch_queue_**max_len** if you update it to be the size of dispatch_queue_?
why not dispatch_queue_len_?

File: src/facade/facade.cc (hunk 2 above)
Comment (dranikpg) [reply src/facade/facade.cc:29]: I assumed it's a very upper worst case limit (if the to assume load on threads is evenly distributed)

File: src/facade/dragonfly_connection.cc [hunk 4]
```diff
@@ -1573,15 +1570,19 @@ void Connection::SendAsync(MessageHandle msg) {
...
     auto it = dispatch_q_.begin();
-    while (it < dispatch_q_.end() && it->IsIntrusive())
+    while (it < dispatch_q_.end() && it->IsControl())
       ++it;
     dispatch_q_.insert(it, std::move(msg));
   } else {
     dispatch_q_.push_back(std::move(msg));
   }
 
+  if (dispatch_q_.size() > stats_->dispatch_queue_max_len) {
```
Comment (romange) [inline src/facade/dragonfly_connection.cc:1582]: @adiholden  here I make it "max"

File: src/facade/facade.cc (hunk 2 above)
Comment (adiholden) [reply src/facade/facade.cc:29]: What do we want to gain from this metric? this will tell us if we want to use ADD or MAX

**Review by adiholden — APPROVED** [review]

**Pull request merged**